odl/test/space/npy_tensors_test.py

```diff
@@ -0,0 +1,1309 @@
+    """Test the inner method against numpy.vdot."""
+    xd = noise_element(tspace)
+    yd = noise_element(tspace)
+
+    # TODO: add weighting
+    correct_inner = np.vdot(yd, xd)
+    assert tspace.inner(xd, yd) == pytest.approx(correct_inner)
+    assert xd.inner(yd) == pytest.approx(correct_inner)
+
+
+def test_inner_exceptions(tspace):
+    """Test if inner raises correctly for bad input."""
+    other_space = odl.rn((4, 3))
+    other_x = other_space.zero()
+    x = tspace.zero()
+
+    with pytest.raises(LinearSpaceTypeError):
+        tspace.inner(other_x, x)
+
+    with pytest.raises(LinearSpaceTypeError):
+        tspace.inner(x, other_x)
+
+
+def test_norm(tspace):
+    """Test the norm method against numpy.linalg.norm."""
+    xarr, x = noise_elements(tspace)
+
+    correct_norm = np.linalg.norm(xarr.ravel())
+    assert tspace.norm(x) == pytest.approx(correct_norm)
+    assert x.norm() == pytest.approx(correct_norm)
+
+
+def test_norm_exceptions(tspace):
+    """Test if norm raises correctly for bad input."""
+    other_space = odl.rn((4, 3))
+    other_x = other_space.zero()
+
+    with pytest.raises(LinearSpaceTypeError):
+        tspace.norm(other_x)
+
+
+def test_pnorm(exponent):
+    """Test the norm method with p!=2 against numpy.linalg.norm."""
+    for tspace in (odl.rn((3, 4), exponent=exponent),
+                   odl.cn((3, 4), exponent=exponent)):
+        xarr, x = noise_elements(tspace)
+        correct_norm = np.linalg.norm(xarr.ravel(), ord=exponent)
+
+        assert tspace.norm(x) == pytest.approx(correct_norm)
+        assert x.norm() == pytest.approx(correct_norm)
+
+
+def test_dist(tspace):
+    """Test the dist method against numpy.linalg.norm of the difference."""
+    [xarr, yarr], [x, y] = noise_elements(tspace, n=2)
+
+    correct_dist = np.linalg.norm((xarr - yarr).ravel())
+    assert tspace.dist(x, y) == pytest.approx(correct_dist)
+    assert x.dist(y) == pytest.approx(correct_dist)
+
+
+def test_dist_exceptions(tspace):
+    """Test if dist raises correctly for bad input."""
+    other_space = odl.rn((4, 3))
+    other_x = other_space.zero()
+    x = tspace.zero()
+
+    with pytest.raises(LinearSpaceTypeError):
+        tspace.dist(other_x, x)
+
+    with pytest.raises(LinearSpaceTypeError):
+        tspace.dist(x, other_x)
+
+
+def test_pdist(exponent):
+    """Test the dist method with p!=2 against numpy.linalg.norm of diff."""
+    for space in (odl.rn((3, 4), exponent=exponent),
+                  odl.cn((3, 4), exponent=exponent)):
+        [xarr, yarr], [x, y] = noise_elements(space, n=2)
+
+        correct_dist = np.linalg.norm((xarr - yarr).ravel(), ord=exponent)
+        assert space.dist(x, y) == pytest.approx(correct_dist)
+        assert x.dist(y) == pytest.approx(correct_dist)
+
+
+def test_space_getitem(getitem_indices):
+    """Check if space indexing works as expected."""
+    space = odl.tensor_space((2, 3, 4), dtype=complex, exponent=1, weighting=2)
+
+    # Ellipsis not supported
```

Comment:
The thing is that there's not really anything interesting to index here. For instance, `space[:2, :]` and `space[1:3, :]` are exactly the same since there's no information attached to the points along the axes. That's different for grids and also product spaces.
Here a utility that condenses an index expression into a new shape would do the deal.